### DISCUSSION TITLE
How can the flux error related to conversion webhook be resolved?

### DISCUSSION THREAD

**Discussion by LittleWat** [fluxcd/flux2#3105] [Q&A]
Hello! I am new to flux and have a question related to slack🧵  https://cloud-native.slack.com/archives/CLAJ40HV3/p1662715738152749

env:

- my local m1 macbook
```sh
$ flux version
flux: v0.32.0
helm-controller: v0.22.2
image-automation-controller: v0.23.5
image-reflector-controller: v0.19.2
kustomize-controller: v0.26.3
notification-controller: v0.24.1
source-controller: v0.25.10

$ minikube version
minikube version: v1.26.0
commit: f4b412861bb746be73053c9f6d2895f12cf78565

$  kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.24.4
Kustomize Version: v4.5.4
Server Version: v1.21.11
````

- On EKS, we met the same error.
```
$ kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.24.4
Kustomize Version: v4.5.4
Server Version: v1.21.14-eks-18ef993
```

Our setting is as follows:

1. Deployed our `v1alpha1` custom resource and `v1alpha1` operator using Flux.
2. Upgraded the operator from `v1alpha1` to `v1alpha2`. The operator implements the conversion webhook following [Implementing conversion - The Kubebuilder Book](https://book.kubebuilder.io/multiversion-tutorial/conversion.html)
3. `kubectl get <resource>.v1alpha2.<foo.com> -n <my-resource-ns> <my-v1alpha1-resource> -o yaml` returns the expected converted(`v1alpha2`) result.
However, flux dry-run continued failing with the following error message.

```log
dry-run failed, error: failed to prune fields: failed add back owned items: failed to convert pruned object at version <foo.com>/v1alpha1: conversion webhook for <foo.com>/v1alpha2, Kind=<resource>  returned invalid metadata: invalid metadata of type <nil> in input object
```

Running the dry-run command seems to be okay:

```sh
$ kubectl apply -f <my-resource>.yaml --dry-run=client
<foo.com>/<my-resource>  configured (dry run)
```

Which part should I check…? Any ideas are really appreciated. Thanks in advance!

**Comment by stefanprodan** [comment]
Flux performs a server-side apply dry-run, it also removes all fields previously applied by kubectl. To simulate some of the things FLux does at apply time do `kubectl apply --server-side --dry-run=server -f`.

**Reply by LittleWat** [reply]
@stefanprodan Thank you for your quick and nice reply!!

In our conversion webhook, metadata is copied as follows:
similar example: https://github.com/IBM/operator-sample-go/blob/b79e66026a5cc5b4994222f2ef7aa962de9f7766/operator-application/api/v1alpha1/application_conversion.go#L37


I run the command and got the following error. How should this be resolved?

Should we NOT copy whole the ObjectMeta (like `dst.ObjectMeta = src.ObjectMeta` )
but should copy the ObjectMeta except for `.metadata.labels.kustomize.toolkit.fluxcd.io/*` ?

```sh
$  kubectl apply --server-side --dry-run=server -f <my-v1alpha1-resource>.yaml 
error: Apply failed with 2 conflicts: conflicts with "kustomize-controller":
- .metadata.labels.kustomize.toolkit.fluxcd.io/name
- .metadata.labels.kustomize.toolkit.fluxcd.io/namespace
Please review the fields above--they currently have other managers. Here
are the ways you can resolve this warning:
* If you intend to manage all of these fields, please re-run the apply
  command with the `--force-conflicts` flag.
* If you do not intend to manage all of the fields, please edit your
  manifest to remove references to the fields that should keep their
  current managers.
* You may co-own fields by updating your manifest to match the existing
  value; in this case, you'll become the manager if the other manager(s)
  stop managing the field (remove it from their configuration).
See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts
```

**Reply by stefanprodan** [reply]
You can pass `--field-manager kustomize-controller`

**Reply by LittleWat** [reply]
Thank you for telling me about the option. (Sorry I am also new to k8s and did not know that option.)

When I pass `--field-manager kustomize-controller`, I got the same error in flux.

```
$  kubectl apply --server-side --dry-run=server -f <my-v1alpha1-resource>.yaml --field-manager kustomize-controller
Error from server: error: failed to prune fields: failed add back owned items: failed to convert pruned object at version <foo.com>/v1alpha1: conversion webhook for <foo.com>/v1alpha2, Kind=<resource>  returned invalid metadata: invalid metadata of type <nil> in input object
```

What point should I check next...? Sorry for the many questions. 🙇

**Reply by stefanprodan** [reply]
Ask the controller runtime folks, IMO your conversion doesn't deal with managed fields and it should, removing metadata during conversion is not OK.

**Reply by LittleWat** [reply]
Thank you for your reply!!

Okay, I will ask the controller runtime folks.

>IMO your conversion doesn't deal with managed fields and it should, removing metadata during conversion is not OK.

I think our conversion(`dst.ObjectMeta = src.ObjectMeta`) deals with `managedFields`

because `metav1.ObjectMeta` includes metadata as follows.
https://github.com/kubernetes/apimachinery/blob/b85d88926e14f20152645d6b697d63dda7929f29/pkg/apis/meta/v1/types.go#L270

Here is the log of `dst` object and `managedFields` seems to be copied from `src`(v1alpha1) to `dst`(v1alpha2).

```json
{
   "dst":{
      "kind":"<MyResource>",
      "apiVersion":"<foo.com>/v1alpha2",
      "metadata":{
         "name":"<name>",
         "namespace":"<ns>",
         "uid":"cc951b88-2263-4f44-abe3-6db1eea02cea",
         "resourceVersion":"110845",
         "generation":10,
         "creationTimestamp":"2022-09-12T06:13:20Z",
         "labels":{
            "kustomize.toolkit.fluxcd.io/name":"<my-sample>",
            "kustomize.toolkit.fluxcd.io/namespace":"flux-system"
         },
         "managedFields":[
            {
               "manager":"kustomize-controller",
               "operation":"Apply",
               "apiVersion":"<foo.com>/v1alpha1",
               "time":"2022-09-12T12:12:51Z",
               "fieldsType":"FieldsV1",
               "fieldsV1":{
                  "f:metadata":{
                     "f:labels":{
                        "f:kustomize.toolkit.fluxcd.io/name":{
                           
                        },
                        "f:kustomize.toolkit.fluxcd.io/namespace":{
                           
                        }
                     }
                  },
                  "f:spec":{
                     ...
                  }
               }
            },
            {
               "manager":"image.binary",
               "operation":"Update",
               "apiVersion":"<foo.com>/v1alpha2",
               "time":"2022-09-12T06:13:21Z",
               "fieldsType":"FieldsV1",
               "fieldsV1":{
                  "f:status":{
                     ...
                  }
               }
            }
         ]
      },
      "spec":{
       ...
      },
      "status":{
         ...
      }
   }
}
```

**Reply by stefanprodan** [reply]
Are you removing the managedFields which no longer map fields in new version?

**Reply by LittleWat** [reply]
I think no. I am just copying the `metav1.ObjectMeta` from src(v1alpha1) to dst(v1alpha2).

Do I need to remove the `managedFields` when we convert?

**Reply by LittleWat** [reply]
`flux logs -f ` shows as follows:

```log
2022-09-14T04:13:36.522Z info Kustomization/my-sample.flux-system - server-side apply completed
2022-09-14T04:13:37.283Z info Kustomization/my-sample.flux-system - server-side apply completed
2022-09-14T04:13:37.351Z info Kustomization/my-sample.flux-system - Reconciliation finished in 3.103322918s, next run in 1m0s
2022-09-14T04:14:39.604Z info Kustomization/my-sample.flux-system - server-side apply completed
2022-09-14T04:14:39.694Z error Kustomization/my-sample.flux-system - Reconciliation failed after 2.340467709s, next try in 1m0s <resource> dry-run failed,  error: failed to prune fields: failed add back owned items: failed to convert pruned object at version <foo.com>/v1alpha1: conversion webhook for <foo.com>/v1alpha2, Kind=<resource>  returned invalid metadata: invalid metadata of type <nil> in input object
2022-09-14T04:15:41.987Z info Kustomization/my-sample.flux-system - server-side apply completed
2022-09-14T04:15:42.714Z info Kustomization/my-sample.flux-system - server-side apply completed
2022-09-14T04:15:42.782Z info Kustomization/my-sample.flux-system - Reconciliation finished in 3.07930296s, next run in 1m0s
```

Almost all the time, it says `server-side apply completed` but sometimes `Reconciliation failed`.

Which part should we dig in?

**Reply by LittleWat** [reply]
From the operator log, I found that `controller-runtime.webhook.webhooks` for validation is not called when it fails. Why `controller-runtime.webhook.webhooks`  for validation  is not sometimes called… 🤔

**Reply by LittleWat** [reply]
I asked #kubebuilder in Kubernetes workspace. Thank you for your help!
https://kubernetes.slack.com/archives/CAR30FCJZ/p1663134731891129

**Reply by LittleWat** [reply]
@stefanprodan Sorry for posting many times 🙇 

First I deployed `v1alpha1` resource manually,
```
$ kubectl apply --server-side -f <manual_v1alpha1>.yaml   
```

Then I repeated this command and it never fails on my local MacBook.
```
$ kubectl apply --server-side --dry-run=server -f <manual_v1alpha1>.yaml  --field-manager kustomize-controller
```

However,  this dry run sometimes fails with the error ` invalid metadata of type <nil> in input object` for `v1alpha1` resource deployed by fluxcd.

So I think the metadata of the resource deployed by fluxcd causes this error. 

I am so sorry to bother you again but could you check the metadata and tell me what I should do?

The metadata of manual one and fluxed one is as follows:
(I got the yaml by running the following command)
```
 kubectl get <my-resource> -n <ns> <name> -o yaml --show-managed-fields
```

- manual_v1alpha1.yaml
```manual_v1alpha1.yaml
apiVersion: <foo.com>/v1alpha2
kind: <MyResource>
metadata:
  creationTimestamp: "2022-09-15T04:19:16Z"
  generation: 1
  managedFields:
  - apiVersion: <foo.com>/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        f:attribute1: {}
        f:attribute2: {}
    manager: kubectl
    operation: Apply
    time: "2022-09-15T04:25:06Z"
  - apiVersion: <foo.com>/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        .: {}
        f:attribute1: {}
        f:attribute2: {}
    manager: before-first-apply
    operation: Update
    time: "2022-09-15T04:25:06Z"
  - apiVersion: <foo.com>/v1alpha2
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:attribute1: {}
    manager: manager
    operation: Update
    time: "2022-09-15T04:44:45Z"
  name: v1alpha1-manual
  namespace: manual
  resourceVersion: "481701"
  uid: 03272b24-503b-480a-8bf6-e33171f00446
spec:
  ...
status:
  ...
``` 


- flux_v1alpha1.yaml
```flux_v1alpha1.yaml
apiVersion: <foo.com>/v1alpha2
kind: <MyResource>
metadata:
  creationTimestamp: "2022-09-15T04:52:03Z"
  generation: 1
  labels:
    kustomize.toolkit.fluxcd.io/name: operator-sample
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  managedFields:
  - apiVersion: <foo.com>/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:kustomize.toolkit.fluxcd.io/name: {}
          f:kustomize.toolkit.fluxcd.io/namespace: {}
      f:spec:
        f:attribute1: {}
        f:attribute2: {}
    manager: kustomize-controller
    operation: Apply
    time: "2022-09-15T04:52:03Z"
  - apiVersion: <foo.com>/v1alpha2
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:attribute1: {}
    manager: manager
    operation: Update
    time: "2022-09-15T04:52:04Z"
  name: v1alpha1-flux
  namespace: flux
  resourceVersion: "483157"
  uid: 696bed77-a12b-45d0-b240-8d685cf790e0

spec:
  ...
status:
  ...
```

**Reply by LittleWat** [reply]
Sorry, this is maybe because there is a bug in the operator. I will fix that and will report the result later.

**Reply by LittleWat** [reply]
I fixed that bug but it did not work... There might be another bug but I could not still find it yet...
I also asked this question in the structured-merge-diff repo
https://github.com/kubernetes-sigs/structured-merge-diff/issues/227

**Reply by LittleWat** [reply]
This seems to be a bug, so I create an issue for this.
https://github.com/fluxcd/flux2/issues/3179

I create a sample repo to reproduce this error.
https://github.com/LittleWat/conversion-webhook-test-with-flux

I am glad if this repo is useful for debugging. Thank you!
